libsodium-bindings/src/LibSodium/Bindings/AEAD.hs

```diff
@@ -0,0 +1,134 @@
+    -> CULLong
+    -- ^ Additional data length.
+    -> Ptr CUChar
+    -- ^ Public nonce of size crypto_aead_xchacha20poly1305_ietf_NPUBBYTES. 
+    -- Should never be reused with the same key. Nonces can be generated using randombytes_buf().
+    -> Ptr CUChar
+    -- ^ Secret key of size crypto_aead_xchacha20poly1305_ietf_KEYBYTES.
+    -> IO CInt
+    -- ^ Returns -1 on failure, 0 on success.
+   
+-- | This is the "detached" version of the encryption function.
+-- The encrypted message and authentication tag are output to different buffers
+-- instead of the tag being appended to the encrypted message.
+foreign import capi "sodium.h crypto_aead_xchacha20poly1305_ietf_encrypt_detached"
+  cryptoAEADXChaCha20Poly1305IETFEncryptDetached
+    :: Ptr CUChar
+    -- ^ Output buffer. Contains the encrypted message with length equal to the message.
+    -> Ptr CUChar
+    -- ^ The authentication tag. Has length crypto_aead_xchacha20poly1305_ietf_ABYTES.
+    -> Ptr CULLong
+    -- ^ Length of the authentication tag buffer.
+    -> Ptr CUChar
+    -- ^ Message to be encrypted.
+    -> CULLong
+    -- ^ Length of input message.
+    -> Ptr CUChar
+    -- ^ Additional, non-confidential data.
+    -> CULLong
+    -- ^ Length of the additional, non-confidential data.
+    -> Ptr CUChar
+    -- ^ Not used in this particular construction, should always be NULL.
+    -> Ptr CUChar
+    -- ^ Public nonce of size crypto_aead_xchacha20poly1305_ietf_NPUBBYTES. 
+    -- Should never be reused with the same key. Nonces can be generated using randombytes_buf().
+    -> Ptr CUChar
+    -- ^ Secret key of size crypto_aead_xchacha20poly1305_ietf_KEYBYTES.
+    -> IO CInt
+    -- ^ Returns -1 on failure, 0 on success.
+
+-- | This is the "detached" version of the decryption function.
+-- Verifies that the authentication tag is valid for the ciphertext, key, nonce,
+-- and additional data.
+foreign import capi "sodium.h crypto_aead_xchacha20poly1305_ietf_decrypt_detached"
+  cryptoAEADXChaCha20Poly1305IETFDecryptDetached
+    :: Ptr CUChar
+    -- ^ If the tag is valid, the ciphertext is decrypted and put into this buffer.
+    -> Ptr CUChar
+    -- ^ Not used in this particular construction, should always be NULL.
+    -> Ptr CUChar
+    -- ^ Ciphertext to be decrypted.
+    -> CULLong
+    -- ^ Length of the ciphertext.
+    -> Ptr CUChar
+    -- ^ The authentication tag. Has length crypto_aead_xchacha20poly1305_ietf_ABYTES.
+    -> Ptr CUChar
+    -- ^ Additional, non-confidential data.
+    -> CULLong 
+    -- ^ Length of the additional, non-confidential data.
+    -> Ptr CUChar
+    -- ^ Public nonce of size crypto_aead_xchacha20poly1305_ietf_NPUBBYTES. 
+    -- Should never be reused with the same key. Nonces can be generated using randombytes_buf().
+    -> Ptr CUChar
+    -- ^ Secret key of size crypto_aead_xchacha20poly1305_ietf_KEYBYTES.
+    -> IO CInt
+    -- Returns 0 on success, -1 if tag is not valid.
+
+-- * Constants.
```

Comment:
I'd recommend you split the sections in the explicit export list of the module. :)